### PR TITLE
Adding GUFunc order, dtype and casting kwarg support

### PR DESCRIPTION
**Pull request by emcastillo**
This mostly completes the support for GUFunc adding the same core characteristics than numpy

### CODE REVIEW

File: cupy/_core/_gufuncs.py [hunk 1]
```diff
@@ -141,6 +142,167 @@ def _validate_normalize_axes(
...
+                    'is not consistent with general signature')
+            if len(outs) != self._nout:
+                raise ValueError(
+                    f'signature {sig} for dtypes is invalid number of inputs '
+                    'is not consistent with general signature')
+
+            ops.append(_OpsRegister._Op(ins, outs, op))
+        return ops
+
+    def _can_cast(self, d1, d2, casting):
```
Comment (asi1024) [inline cupy/_core/_gufuncs.py:201]: In what case does this function behave differently from `numpy.can_cast` itself?

File: cupy/_core/_gufuncs.py [hunk 2]
```diff
@@ -141,6 +142,167 @@ def _validate_normalize_axes(
...
+
+    def _sig_str_to_tuple(self, sig):
+        sig = sig.replace(' ', '')
+        toks = sig.split('->')
+        if sig.startswith('->'):
+            ins = ''
+            outs = toks
+        elif sig.endswith('->'):
+            ins = toks
+            outs = ''
```
[the remark this answers is not in the document]
Comment (emcastillo) [reply cupy/_core/_gufuncs.py:173]: yeah, sorry about m stupidity here

File: cupy/_core/_gufuncs.py (hunk 1 above)
Comment (emcastillo) [reply cupy/_core/_gufuncs.py:201]: Its the same but with a shortcut, this was taken from kernel.pyx iirc

**Comment by emcastillo** [comment]
Added tests for both order and signatures and fixed some quirks with # of outputs.

PTAL!

I will add tests for the rest of features (I did it before but launched a git clean before adding the file :() and expose gufunc to the user via cupyx in a different PR.

**Comment by asi1024** [comment]
Jenkins, test this please.

**Comment by emcastillo** [comment]
failure is not related :)

**Comment by asi1024** [comment]
I confirmed that `test_gufunc.py` ran without errors in windows CI. LGTM.

**Comment by emcastillo** [comment]
Thank you!

**Pull request merged**